File: day-2/part-2.cc
#include <fstream>
#include <iostream>

int main(int argc, char* argv[]) {
  std::ifstream input_file(argv[1]);

  std::string direction;
  int units;

  int horiz = 0, depth = 0, aim = 0;

  while (input_file >> direction >> units) {
    if (direction == "forward") {
      horiz += units;
      depth += aim * units;
    } else if (direction == "down") {
      aim += units;
    } else if (direction == "up") {
      aim -= units;
    }
  }

  std::cout << horiz * depth << "\n";
}

File: day-2/part-2.dart
import 'dart:convert';
import 'dart:io';

void main(List<String> args) async {
  var file = File(args[0]);

  var horiz = 0, depth = 0, aim = 0;

  await file
    .openRead()
    .transform(utf8.decoder)
    .transform(new LineSplitter())
    .forEach((line) {
      var dir_and_unit = line.split(' ');
      var dir = dir_and_unit[0];
      var unit = int.parse(dir_and_unit[1]);

      switch(dir) {
        case 'forward':
          horiz += unit;
          depth += aim * unit;
          break;
        case 'down': aim += unit; break;
        case 'up': aim -= unit; break;
      }
    });

  print(horiz * depth);
}
